src/Features/CSharp/Portable/CodeRefactorings/ConvertIfToSwitch/CSharpConvertIfToSwitchCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,285 @@
+                                return null;
+                        }
+
+                        var p = CreatePatternFromExpression(leftmost, semanticModel, ref switchExpression);
+                        if (p == null)
+                        {
+                            return null;
+                        }
+
+                        // In case of chained "&&" operators, we need to justify the precedence of operands in which
+                        // the "&&" operator is applied. For example, in "if (expr && cond1 && cond2)" we need to
+                        // separate out the leftmost expression and reconstruct the rest of conditions into a new 
+                        // expression, so that they would be evaluated entirely after "expr" e.g. "expr && (cond1 && cond2)".
+                        // Afterwards, we can directly use it in a case guard e.g. "case pat when cond1 && cond2:".
+                        var condition = (leftmost.Parent.Parent as BinaryExpressionSyntax)
+                            ?.WithLeft(((BinaryExpressionSyntax)leftmost.Parent).Right);
+
+                        return new Pattern.Guarded(p, (condition ?? node.Right).WalkDownParentheses());
+                    }
+
+                default:
+                    return null;
+            }
+        }
+
+        private static Task<Document> UpdateDocumentAsync(
+            SyntaxNode root,
+            Document document,
+            StatementSyntax ifStatement,
+            StatementSyntax elseBody,
+            ExpressionSyntax switchExpression,
+            SemanticModel semanticModel,
+            List<(List<Pattern> patterns, StatementSyntax body)> sectionData)
+        {
+            var sections = sectionData.Select(data => SwitchSection(data.body, semanticModel,
+                labels: SyntaxFactory.List<SwitchLabelSyntax>(data.patterns.Select(p => p.CreateSwitchLabel()))));
+
+            var sectionList = SyntaxFactory.List(sections);
+
+            if (elseBody != null)
+            {
+                sectionList = sectionList.Add(SwitchSection(elseBody, semanticModel,
+                    labels: SyntaxFactory.SingletonList<SwitchLabelSyntax>(SyntaxFactory.DefaultSwitchLabel())));
+            }
+
+            var @switch = SyntaxFactory.SwitchStatement(switchExpression, sectionList);
+            var newRoot = root.ReplaceNode(ifStatement, @switch);
+            return Task.FromResult(document.WithSyntaxRoot(newRoot));
+        }
+
+        private static SwitchSectionSyntax SwitchSection(StatementSyntax body, SemanticModel semanticModel, SyntaxList<SwitchLabelSyntax> labels)
+             => SyntaxFactory.SwitchSection(labels, SyntaxFactory.List(GetSwitchSectionBody(body, semanticModel)));
+
+        private static IEnumerable<SyntaxNode> GetSwitchSectionBody(StatementSyntax node, SemanticModel semanticModel)
+        {
+            var requiresBreak = semanticModel.AnalyzeControlFlow(node).EndPointIsReachable;
+
```

Comment:
@CyrusNajmabadi  This could be a local function to be lazily evaluated, because in case of an empty block it is definitely true and we don't need to check. Is that ok here to use a local function (for example)?